src/puya/awst/validation/immutable.py

```diff
@@ -0,0 +1,33 @@
+from puya import log
+from puya.awst import nodes as awst_nodes
+from puya.awst.awst_traverser import AWSTTraverser
+
+logger = log.get_logger(__name__)
+
+
+class ImmutableValidator(AWSTTraverser):
```

Comment:
Should this include mutating nodes like `ArrayExtend` and `ArrayPop`?

Reply:
Good point, initially this issue was addressing arc4.Address, which those nodes don't operate on. But we should still check regardless.